File: js/convert-sorted-array-to-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
// Time complexity: O(n) where n is the length of the input array
// Space complexity: O(n) if we count the result, otherwise O(log n) for the recursive call stack
// recursive
var sortedArrayToBST = function (nums) {
    const dfs = (left, right) => {
        if (left > right)
            return null;

        const mid = Math.floor((right - left) / 2) + left
        const node = new TreeNode(nums[mid])

        node.left = dfs(left, mid - 1)
        node.right = dfs(mid + 1, right)

        return node
    }

    return dfs(0, nums.length - 1)
};

File: python/convert-sorted-array-to-binary-search-tree.py
from typing import List, Optional


class TreeNode:
    # Definition for a binary tree node.
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # Time complexity: O(n) where n is thee length of the input array
    # Space complexity: O(n) if we count the result, otherwise O(log n) for the call stack
    # recursive
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def dfs(left, right):
            if left > right:
                return None

            mid = ((right - left)//2)+left

            node = TreeNode(nums[mid])

            node.left = dfs(left, mid-1)
            node.right = dfs(mid+1, right)

            return node

        return dfs(0, len(nums)-1)
